### PR TITLE
More robust JSON prompting and parsing

### PR DESCRIPTION
**Pull request by mrtj**
This PR implements a langchain-style, hopefully more robust output parsing as discussed in #761 .

### CODE REVIEW

**Comment by mrtj** [comment]
Currently I added a `use_langchain_parser` configuration attribute to each supported metrics. Settings this to `true` will use the JSON output parser implementation directly from LangChain, but in this case the retry mechanism is not (yet) implemented. Setting this to `false` will use the original JSON parsing from ragas, and verify the extracted JSON object with pydantic. 

In both cases the output instructions and the expected JSON schema is injected in the prompts, and the models tend to follow correctly these instructions and generate a JSON output conforming to the schema.

I've tested the behavior with gpt-3.5-turbo, gpt-4-turbo, Claude v2, v2.1, v3 Sonnet and Claude Instant v1 (via Amazon Bedrock), and each models usually return a correct JSON format.

**Comment by shahules786** [comment]
Hey @mrtj perfect timing. I was collecting issues related to the same and thinking of your work + how I can help. 
I'll take a look at this today. Thanks a lot :)

**Comment by shahules786** [comment]
Just took a detailed look at it. Great work, exactly what we needed. Some todo's for me

- [ ] Implement RetryWithErrorOutputParser
- [ ] Remove `json_load` if not required
- [ ] Convert all prompt (testset) outputs to same pydantic model
- [ ] Make sure prompt adaptation is working 
- [ ] Minor formating changes

**Comment by shahules786** [comment]
Hey @mrtj I have made some changes. would love to have your opinion on it as well if any.

**Review by shahules786 — APPROVED** [review]

**Comment by mrtj** [comment]
Refactoring the try..catch blocks around the output parsing is an idea that I was hesitating how to implement the best: having a custom subclass of pydantic output parser is certainly the most elegant way, I like it very much.

The other thing I wanted to ask: could you explain me why do you think that asking a binary integer (0/1) response from the model is better than using a boolean directly? My experience with LLMs suggest that usually it is a good idea to restrict the grade of liberty of the models about the response format. This is the reason I've converted these variables to boolean. But maybe you have some other reasons?

**Comment by shahules786** [comment]
Hey @mrtj thank you.

The reason for converting verdicts to boolean is to make it language agnostic - ie, during prompt adaptation we convert a few shot examples to the target language. So in this case the verdict should stay the same irrespective of the language.
https://docs.ragas.io/en/stable/howtos/applications/use_prompt_adaptation.html

Any better idea to do the same?

File: src/ragas/llms/prompt.py [hunk 1]
```diff
@@ -136,7 +147,7 @@ def get_example_str(self, example_no: int) -> str:
                 else value
             )
             example_str += f"\n{key}: {value}"
-        return example_str
+        return "```" + example_str + "```"
```
Comment (jjmachan) [inline src/ragas/llms/prompt.py:150]: why do we use the triple backticks ?
other models seem to mess this up by adding extra tokens like ```bash etc

**Review by jjmachan — APPROVED** [review]
hey @mrtj thanks a lot for this PR ❤️ 

I tested this will a few other OSS models and Claude and there are improvements there as well. We will have to write some tests for JSON parsing across OSS models I think so that we keep track of model changes and their support for JSON. I'll work on those


But this improves the major issue the community are facing today and Thanks a lot for contributing this 🙂

**Comment by larunach** [comment]
> Currently I added a `use_langchain_parser` configuration attribute to each supported metrics. Setting this to `true` will use the JSON output parser implementation directly from LangChain, but in this case the retry mechanism is not (yet) implemented. Setting this to `false` will use the original JSON parsing from ragas, and verify the extracted JSON object with pydantic.
> 
> In both cases the output instructions and the expected JSON schema are injected in the prompts, and the models tend to follow correctly these instructions and generate a JSON output conforming to the schema.
> 
> I've tested the behavior with gpt-3.5-turbo, gpt-4-turbo, Claude v2, v2.1, v3 Sonnet and Claude Instant v1 (via Amazon Bedrock), and each models usually return a correct JSON format.

I am using mistral model via hugging face and it returns output in this format, still it is not able to parse the output
 ```
{
    "reason": "The provided context was useful in arriving at the given answer. The context explicitly states that the capital of India is Delhi, which matches the answer.",
    "verdict": 1
}
```
What is the issue with this  ?

File: src/ragas/llms/prompt.py (hunk 1 above)
Comment (mrtj) [reply src/ragas/llms/prompt.py:150]: if I am not mistaken, the json/pydantic response parser in langchain search for the triple backsticks in the response to identify the json structure in the sometimes more verbose response of the model. Some models love to add some free text preamble before the json block, or conclusion/explanation of their work done after the json.

File: docs/getstarted/evaluation.md [hunk 2]
```diff
@@ -3,10 +3,6 @@
 
 Once your test set is ready (whether you've created your own or used the [synthetic test set generation module](get-started-testset-generation)), it's time to evaluate your RAG pipeline. This guide assists you in setting up Ragas as quickly as possible, enabling you to focus on enhancing your Retrieval Augmented Generation pipelines while this library ensures that your modifications are improving the entire pipeline.
 
-<p align="left">
```
Comment (mtharrison) [inline docs/getstarted/evaluation.md:6]: Just curious, what was the reason for removing the diagram?

**Pull request merged**